### PR TITLE
Add default favicon for favicon.png and favicon.ico

### LINKED ISSUE #33: Default favicon
The issue of web browsers requesting `/favicon.ico` is starting to become confusing.

I would like to suggest that we supply a default favicon icon, compiled into the file server — for example, the Spin logo.

Users could always override that by providing their own file at the root of the files.

Thoughts?

### PR DESCRIPTION
**Pull request by ThorstenHans**
With this PR, `spin-fileserver` is able to serve either `favicon.ico` or `favicon.png` if the user did not provide a corresponding file.

Instead of altering many parts of the existing `FileServer` implementation, I came up with a `ResourceServer` struct responsible for reading and compressing the favicon from fallback favicon graphics embedded into the module.

This also required modifying the return value of `FileServer::resolve` from `Option<PathBuf>` to a custom enum that I called `FileServerPath`. The `FileServerPath` comes with different variants to represent physical resources (`Physical(PathBuf)`), embedded resources (`Embedded(&'static [u8])`)) and non-existing resources (`None`).

fixes #33

### CODE REVIEW

**Comment by ThorstenHans** [comment]
Quick addition: I added `favicon.png` because it would also work in the case someone links the favicon as png without actually providing it.

This is totally optional, removing it would of course reduce the amount of modifications.

**Review by vdice — COMMENTED** [review]
This will be a handy feature, thank you @ThorstenHans!

Found a small bug in my testing, noted in comments.

Additionally, I wanted to bring up the scenario of when an app uses this component but sets a trigger route other than root (`/...`), eg:
```
[component.trigger]
route = "/static/..."
```
With such an app, the favicon fallback logic works as expected when the request is  e.g. `http://localhost:3000/static/favicon.ico` (where `PATH_INFO_HEADER` is relative to the trigger route, so `/favicon.ico`)  -- yet, favicons are always requested at the root of the website, so requests to e.g. `http://localhost:3000/favicon.ico` still encounter a 404.  I wonder what we could do to support favicon fallback when arbitrary trigger routes are used?  

(I stumbled upon this on `spin new` using the spin-fileserver template and cruising through the prompts; the default for the trigger route ("HTTP path") happens to be `/static/...`

File: src/lib.rs [hunk 1]
```diff
@@ -404,4 +463,19 @@ mod tests {
         let rsp = <super::SpinHttp as spin_http::SpinHttp>::handle_http_request(req);
         assert_eq!(rsp.status, 200);
     }
+
+    #[test]
+    fn test_serve_fallback_favicon() {
+        let req = spin_http::Request {
+            method: spin_http::Method::Get,
+            uri: "http://thisistest.com/".to_string(),
+            headers: vec![(PATH_INFO_HEADER.to_string(), "favicon.png".to_string())],
```
[the remark this answers is not in the document]
Comment (ThorstenHans) [reply src/lib.rs]: We definitely want to use the const here 😄 as well and get rid of the magic string

File: src/lib.rs [hunk 2]
```diff
@@ -82,19 +89,55 @@ fn serve(req: Request) -> Result<Response> {
...
+}
+
+trait IsFavicon {
+    fn is_favicon(&self) -> bool;
+}
+
+impl IsFavicon for PathBuf {
+    fn is_favicon(&self) -> bool {
+        match self.clone().into_os_string().into_string() {
+            Ok(s) => s == FAVICON_ICO_FILENAME || s == FAVICON_PNG_FILENAME,
```
[the remark this answers is not in the document]
Comment (ThorstenHans) [reply src/lib.rs]: updated based on your suggestion.

**Comment by ThorstenHans** [comment]
> This will be a handy feature, thank you @ThorstenHans!
> 
> Found a small bug in my testing, noted in comments.
> 
> Additionally, I wanted to bring up the scenario of when an app uses this component but sets a trigger route other than root (`/...`), eg:
> 
> ```
> [component.trigger]
> route = "/static/..."
> ```
> 
> With such an app, the favicon fallback logic works as expected when the request is e.g. `http://localhost:3000/static/favicon.ico` (where `PATH_INFO_HEADER` is relative to the trigger route, so `/favicon.ico`) -- yet, favicons are always requested at the root of the website*, so requests to e.g. `http://localhost:3000/favicon.ico` still encounter a 404. I wonder what we could do to support favicon fallback when arbitrary trigger routes are used?
> 
> __edit: perhaps both `/favicon.*` and `/images/favicon.*` are supported by browsers? Is there a canonical spec where they are expected to live?__
> 
> (I stumbled upon this on `spin new` using the spin-fileserver template and cruising through the prompts; the default for the trigger route ("HTTP path") happens to be `/static/...`

I looked at the HTLM 4.0.1 and XHTML 1.0 spec to spot any further information about favicon requests. I found this - quite dated - blog post also talking about the loading favicons from different sources... However, everything I found assumes that users actually specify a custom favicon. 

According to the HTML5 spec (https://html.spec.whatwg.org/multipage/links.html#linkTypes) using `<link rel="icon" ...>` is treated as [External Resource](https://html.spec.whatwg.org/multipage/links.html#external-resource-link). I quickly read through those pages, but haven't found any **spec** that explains more...

Regarding the with either specifying

```toml
trigger = { type = "http", base = "/foo" }
# ...
route = "/..."
```

or

```toml
trigger = { type = "http", base = "" }
# ...
route = "/bar/..."
```

TBH, I don't know how we could address this when the user doesn't specify any favicon in HTML.  Obviously, it'll serve the fallback icon for both configurations shown before when user specifies 

```html
<link rel="shortcut icon" href="favicon.ico" type="image/x-icon">
``` 
(See `href` not starting with a leading `/`.)

**Comment by vdice** [comment]
Thanks for the research on conventions and preferred methods of specifying a favicon.  Like you've pointed out, it does seem that HTML representation is recommended over varying browser conventions for auto-lookup.

> TBH, I don't know how we could address this when the user doesn't specify any favicon in HTML. Obviously, it'll serve the fallback icon for both configurations shown before when user specifies

Indeed, with how Spin's trigger mechanism is designed, I believe you are right.

**Review by vdice — COMMENTED** [review]
Functionality-wise, this looks and works great for me!

Lastly, could we add a bit of documentation for this feature in the readme.md?

**Comment by ThorstenHans** [comment]
Sure, I'll add a corresponding section to the README. Will look into it tomorrow morning

**Review by vdice — APPROVED** [review]
LGTM!  Thank you @ThorstenHans.

**Pull request merged**